File: vfsimulator/core/program_analysis.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from __future__ import annotations

from typing import Any, Dict, List


class ProgramAnalyzer:
    def __init__(self, params: Dict[str, Any]) -> None:
        self.params = params

    @staticmethod
    def is_vreg_name(name: Any) -> bool:
        return isinstance(name, str) and name.startswith("v") and name[1:].isdigit()

    def resolve_bound(self, bound: Any) -> int:
        if isinstance(bound, int):
            return bound
        if isinstance(bound, str):
            if bound.isdigit():
                return int(bound)
            if bound in self.params:
                return int(self.params[bound])
        raise ValueError(f"Unsupported loop bound: {bound}")

    def resolve_unroll_value(self, unroll: Any) -> int:
        if isinstance(unroll, int):
            return max(1, int(unroll))
        if isinstance(unroll, str):
            if unroll.isdigit():
                return max(1, int(unroll))
            if unroll in self.params:
                return max(1, int(self.params[unroll]))
        return 1

    def iter_insts(self, node: Any):
        if isinstance(node, list):
            for x in node:
                yield from self.iter_insts(x)
            return
        if not isinstance(node, dict):
            return
        if node.get("type") == "inst":
            yield node
        body = node.get("body")
        if isinstance(body, list):
            for x in body:
                yield from self.iter_insts(x)

    def collect_vreg_capacity_warnings(
        self,
        program: List[Dict[str, Any]],
        preg_num: int,
    ) -> List[Dict[str, Any]]:
        warnings: List[Dict[str, Any]] = []

        def walk(nodes: Any, path: str) -> None:
            if not isinstance(nodes, list):
                return
            loop_idx = 0
            for node in nodes:
                if not isinstance(node, dict):
                    continue
                if node.get("type") != "loop":
                    continue

                loop_idx += 1
                loop_path = f"{path}.loop{loop_idx}"
                unroll = self.resolve_unroll_value(node.get("unroll", 1))
                body = node.get("body", [])

                vregs = set()
                for inst in self.iter_insts(body):
                    for x in inst.get("src", []) or []:
                        if self.is_vreg_name(x):
                            vregs.add(x)
                    for x in inst.get("dst", []) or []:
                        if self.is_vreg_name(x):
                            vregs.add(x)

                base_vreg_namespace = len(vregs)
                expanded_vreg_namespace = base_vreg_namespace * max(1, unroll)

                if expanded_vreg_namespace > preg_num:
                    warnings.append(
                        {
                            "kind": "vreg_namespace_overflow_risk",
                            "loop_path": loop_path,
                            "preg_num": int(preg_num),
                            "base_vreg_namespace": int(base_vreg_namespace),
                            "unroll": int(unroll),
                            "expanded_vreg_namespace": int(expanded_vreg_namespace),
                            "message": (
                                "Unroll-expanded virtual-register namespace exceeds physical register count. "
                                "Prediction may be low-confidence for this case."
                            ),
                        }
                    )

                walk(body, loop_path)

        walk(program, "program")
        return warnings

    def infer_nested_bounds_from_loop(self, loop_node: Dict[str, Any]) -> List[int]:
        bounds: List[int] = []
        node = loop_node

        while isinstance(node, dict) and node.get("type") == "loop" and len(bounds) < 3:
            bounds.append(self.resolve_bound(node.get("iters")))
            body = node.get("body", [])

            next_loop = None
            if isinstance(body, list):
                for item in body:
                    if isinstance(item, dict) and item.get("type") == "loop":
                        next_loop = item
                        break

            if next_loop is None:
                break
            node = next_loop

        return bounds

    def infer_top_block_loop_bounds(
        self,
        program: List[Dict[str, Any]],
    ) -> Dict[int, List[int]]:
        result: Dict[int, List[int]] = {}
        tbid = 0

        if not isinstance(program, list):
            return {0: []}

        for node in program:
            if isinstance(node, dict) and node.get("type") == "loop":
                result[tbid] = self.infer_nested_bounds_from_loop(node)
                tbid += 1

        if not result:
            result[0] = []

        return result

File: native/ProgramAnalysis.cpp
// Copyright (c) 2026 Huawei Technologies Co., Ltd.
// This program is free software, you can redistribute it and/or modify it under the terms and conditions of
// CANN Open Software License Agreement Version 2.0 (the "License").
// Please refer to the License for details. You may not use this file except in compliance with the License.
// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A PARTICULAR PURPOSE.
// See LICENSE in the root of the software repository for the full text of the License.

#include "native/ProgramAnalysis.h"

#include <algorithm>
#include <cctype>
#include <stdexcept>
#include <utility>

namespace vfsim {
namespace {

bool isDigits(const std::string &text) {
  if (text.empty())
    return false;
  return std::all_of(text.begin(), text.end(), [](unsigned char c) {
    return std::isdigit(c) != 0;
  });
}

} // namespace

ProgramNode ProgramNode::makeInst(ProgramInstNode value) {
  ProgramNode node;
  node.kind = Kind::Inst;
  node.inst = std::move(value);
  node.loop.reset();
  return node;
}

ProgramNode ProgramNode::makeLoop(ProgramLoopNode value) {
  ProgramNode node;
  node.kind = Kind::Loop;
  node.loop = std::make_shared<ProgramLoopNode>(std::move(value));
  return node;
}

ProgramAnalysis::ProgramAnalysis(ParamMap params) : params_(std::move(params)) {}

bool ProgramAnalysis::isVregName(const std::string &name) {
  if (name.size() < 2 || name[0] != 'v')
    return false;
  return isDigits(name.substr(1));
}

int64_t ProgramAnalysis::resolveBound(const std::string &bound) const {
  if (isDigits(bound))
    return std::stoll(bound);
  auto it = params_.find(bound);
  if (it != params_.end())
    return it->second;
  throw std::invalid_argument("Unsupported loop bound: " + bound);
}

int64_t ProgramAnalysis::resolveUnrollValue(const std::string &unroll) const {
  if (isDigits(unroll))
    return std::max<int64_t>(1, std::stoll(unroll));
  auto it = params_.find(unroll);
  if (it != params_.end())
    return std::max<int64_t>(1, it->second);
  return 1;
}

void ProgramAnalysis::collectVregsFromInst(
    const ProgramInstNode &inst,
    std::unordered_map<std::string, bool> &vregs) {
  for (const auto &name : inst.src) {
    if (isVregName(name))
      vregs.emplace(name, true);
  }
  for (const auto &name : inst.dst) {
    if (isVregName(name))
      vregs.emplace(name, true);
  }
}

void ProgramAnalysis::walkVregWarnings(const std::vector<ProgramNode> &nodes,
                                       const std::string &path,
                                       int64_t pregNum,
                                       std::vector<VregCapacityWarning> &warnings) const {
  int loopIndex = 0;
  for (const auto &node : nodes) {
    if (node.kind != ProgramNode::Kind::Loop || !node.loop)
      continue;

    ++loopIndex;
    const ProgramLoopNode &loop = *node.loop;
    const std::string loopPath = path + ".loop" + std::to_string(loopIndex);
    const int64_t unroll = resolveUnrollValue(loop.unroll);

    std::unordered_map<std::string, bool> vregs;
    std::vector<const ProgramNode *> stack;
    for (const auto &child : loop.body)
      stack.push_back(&child);

    while (!stack.empty()) {
      const ProgramNode *cur = stack.back();
      stack.pop_back();
      if (cur->kind == ProgramNode::Kind::Inst) {
        collectVregsFromInst(cur->inst, vregs);
        continue;
      }
      if (cur->loop) {
        for (const auto &child : cur->loop->body)
          stack.push_back(&child);
      }
    }

    const int64_t baseVregNamespace = static_cast<int64_t>(vregs.size());
    const int64_t expandedVregNamespace = baseVregNamespace * std::max<int64_t>(1, unroll);
    if (expandedVregNamespace > pregNum) {
      warnings.push_back(VregCapacityWarning{
          "vreg_namespace_overflow_risk",
          loopPath,
          pregNum,
          baseVregNamespace,
          unroll,
          expandedVregNamespace,
          "Unroll-expanded virtual-register namespace exceeds physical register count. Prediction may be low-confidence for this case.",
      });
    }

    walkVregWarnings(loop.body, loopPath, pregNum, warnings);
  }
}

std::vector<VregCapacityWarning>
ProgramAnalysis::collectVregCapacityWarnings(const std::vector<ProgramNode> &program,
                                             int64_t pregNum) const {
  std::vector<VregCapacityWarning> warnings;
  walkVregWarnings(program, "program", pregNum, warnings);
  return warnings;
}

std::vector<int64_t> ProgramAnalysis::inferNestedBoundsFromLoop(const ProgramLoopNode &loop) const {
  std::vector<int64_t> bounds;
  const ProgramLoopNode *cur = &loop;

  while (cur != nullptr && bounds.size() < 3) {
    bounds.push_back(resolveBound(cur->iters));
    const ProgramLoopNode *nextLoop = nullptr;
    for (const auto &node : cur->body) {
      if (node.kind == ProgramNode::Kind::Loop && node.loop) {
        nextLoop = node.loop.get();
        break;
      }
    }
    cur = nextLoop;
  }

  return bounds;
}

std::unordered_map<int, std::vector<int64_t>>
ProgramAnalysis::inferTopBlockLoopBounds(const std::vector<ProgramNode> &program) const {
  std::unordered_map<int, std::vector<int64_t>> result;
  int tbid = 0;

  for (const auto &node : program) {
    if (node.kind == ProgramNode::Kind::Loop && node.loop) {
      result.emplace(tbid++, inferNestedBoundsFromLoop(*node.loop));
    }
  }

  if (result.empty())
    result.emplace(0, std::vector<int64_t>{});

  return result;
}

} // namespace vfsim
